File: tests/sparse_faces.cpp
/*
 * Olivetty faces is a 400x4096 image dataset useful for quick checking and
 * debugging.
 */


#include <fstream>

#include "../src/nnd.h"

using namespace nndescent;


// Timer for returning passed time in milliseconds.
Timer test_timer;


// Read csv from disk and return as nndescent::Matrix.
Matrix<float> read_csv(std::string file_path)
{
    std::cout << "Reading " << file_path << "\n";

    std::fstream csv_file;
    csv_file.open(file_path, std::ios::in);

    if (!std::ifstream(file_path))
    {
        std::cerr << "Dataset '" << file_path
                << "' not found. Did you try to run 'make_test_data.py'?\n";
        exit(1);
    }

    size_t n_rows = 0;
    std::vector<float> vec_data;
    std::string line;
    while (std::getline(csv_file, line))
    {
        ++n_rows;
        std::vector<float> row;
        std::stringstream ss_line(line);
        while (ss_line.good())
        {
            std::string substr;
            getline(ss_line, substr, ',');
            vec_data.push_back(atof(substr.c_str()));
        }
    }
    csv_file.close();

    Matrix<float> matrix(n_rows, vec_data);

    return matrix;
}


int main()
{
    // DATA INPUT

    // Reset the timer to 0.
    test_timer.start();

    std::string train_path = (std::string)getenv("HOME")
        + "/Downloads/nndescent_test_data/faces_train.csv";

    std::string test_path = (std::string)getenv("HOME")
        + "/Downloads/nndescent_test_data/faces_test.csv";

    // Read data from csv file.
    Matrix<float> data = read_csv(train_path);
    Matrix<float> query_data = read_csv(test_path);

    // Convert data to csr form.
    CSRMatrix<float> csr_data(data);
    CSRMatrix<float> csr_query_data(query_data);

    test_timer.stop("Reading csv");


    // NEAREST NEIGHBORS - TRAINING

    // Define algorithm parameters
    Parms parms;
    parms.n_neighbors=30;
    parms.verbose=true;
    parms.seed=1234;

    test_timer.start();

    // Run nearest neighbor descent algorithm.
    NNDescent nnd = NNDescent(csr_data, parms);

    test_timer.stop("nnd");

    // Return approximate NN-Matrix and distances.
    Matrix<int> nn_indices = nnd.neighbor_indices;
    Matrix<float> nn_distances = nnd.neighbor_distances;

    test_timer.start();

    // Use brute force algorithm for exact values.
    parms.algorithm="bf";
    NNDescent nnd_bf = NNDescent(csr_data, parms);

    test_timer.stop("brute force");

    // Return exact NN-Matrix.
    Matrix<int> nn_indices_ect = nnd_bf.neighbor_indices;
    Matrix<float> nn_distances_ect = nnd_bf.neighbor_distances;

    // Calculate accuracy of NNDescent.
    std::cout << "Accuracy of NN:\n";
    recall_accuracy(nn_indices, nn_indices_ect);


    // NEAREST NEIGHBORS - TESTING

    const int k_query = 10;
    test_timer.start();

    // Calculate k_query-nearest neighbors for each query point
    nnd.query(csr_query_data, k_query);

    test_timer.stop("nnd query");

    // Return approximate NN-Matrix and distances.
    Matrix<int> nn_query_indices = nnd.query_indices;
    Matrix<float> nn_query_distances = nnd.query_distances;

    test_timer.start();

    // Same using brute force algorithm.
    nnd_bf.query(csr_query_data, k_query);

    test_timer.stop("brute force query");

    // Return approximate NN-Matrix and distances.
    Matrix<int> nn_query_indices_ect = nnd_bf.query_indices;
    Matrix<float> nn_query_distances_ect = nnd_bf.query_distances;

    // Calculate accuracy of NNDescent.
    std::cout << "Accuracy of NN query csr_data:\n";
    recall_accuracy(nn_query_indices, nn_query_indices_ect);

    return 0;
}

File: tests/sparse_faces.py
"""
Olivetty faces is a 400x4096 image dataset useful for quick checking
and debugging.
"""

import os

from sklearn.datasets import fetch_olivetti_faces
from sklearn.model_selection import train_test_split
import nndescent
import numpy as np
import scipy.sparse

DATA_PATH = os.path.expanduser("~/Downloads/nndescent_test_data")


def accuracy(approx_neighbors, true_neighbors):
    """Returns accuracy of algorithm when compared with exact values."""
    result = np.zeros(approx_neighbors.shape[0])
    for i in range(approx_neighbors.shape[0]):
        n_correct = np.intersect1d(
            approx_neighbors[i], true_neighbors[i]
        ).shape[0]
        result[i] = n_correct / true_neighbors.shape[1]
    print(f"Average accuracy of {np.mean(result)}\n")


def nn_brute_force(train, test, n_neighbors):
    """Calculates nearest neighbors (of test in train) using brute force."""
    nn_indices_test = []
    nn_distances_test = []
    for pnt_test in test:
        distances = [
            [i, np.linalg.norm(pnt_test - pnt_train)]
            for i, pnt_train in enumerate(train)
        ]
        distances.sort(key=lambda x: x[1])
        nn_indices_test.append([x for x, _ in distances[:n_neighbors]])
        nn_distances_test.append([y for _, y in distances[:n_neighbors]])
    return np.array(nn_indices_test), np.array(nn_distances_test)


# Data
olivetti_faces = fetch_olivetti_faces(data_home=DATA_PATH).data
data, query_data = train_test_split(
    olivetti_faces, test_size=0.2, random_state=1234
)

# Convert to sparse format
csr_data = scipy.sparse.csr_matrix(data)
csr_query_data = scipy.sparse.csr_matrix(query_data)

# NEAREST NEIGHBORS - TRAINING

# Run NND algorithm.
n_neighbors = 30
nnd = nndescent.NNDescent(csr_data, n_neighbors=n_neighbors)

# Get result
nn_indices, nn_distances = nnd.neighbor_graph

# Calculate exact nearest neighbors for comparison.
nn_indices_ect, nn_distances_ect = nn_brute_force(data, data, n_neighbors)

print("\nInput csr_data\n", csr_data)
print("\nApproximate nearest neighbors\n", nn_indices)
print("\nExact nearest neighbors\n", nn_indices_ect)
print("\nApproximate nearest neighbors distances\n", nn_distances)
print("\nExact nearest neighbors distances\n", nn_distances_ect)
print("\nAccuracy of nndescent algorithm compared with exact values (train):")
accuracy(nn_indices, nn_indices_ect)


# NEAREST NEIGHBORS - TESTING

# Calculate nearest neighbors for each query point.
k_query = 10
nn_query_indices, nn_query_distances = nnd.query(csr_query_data, k=k_query)

# Calculate exact nearest neighbors for comparison.
nn_query_indices_ect, nn_query_distances_ect = nn_brute_force(data, query_data, k_query)


print("\nInput query csr_data\n", csr_data)
print("\nApproximate nearest neighbors of query points\n", nn_query_indices)
print("\nExact nearest neighbors of query points\n", nn_query_indices_ect)
print("\nApproximate nearest neighbors query distances\n", nn_query_distances)
print("\nExact nearest neighbors query distances\n", nn_query_distances_ect)
print("\nAccuracy of nndescent algorithm compared with exact values (test):")
accuracy(nn_query_indices, nn_query_indices_ect)
